src/main/java/com/fasterxml/jackson/databind/util/clhm/ConcurrentLinkedHashMap.java

```diff
@@ -0,0 +1,1597 @@
+        final Node<K, V> node = data.get(key);
+        if (node == null) {
+            return null;
+        }
+        afterRead(node);
+        return node.getValue();
+    }
+
+    /**
+     * Returns the value to which the specified key is mapped, or {@code null}
+     * if this map contains no mapping for the key. This method differs from
+     * {@link #get(Object)} in that it does not record the operation with the
+     * page replacement policy.
+     *
+     * @param key the key whose associated value is to be returned
+     * @return the value to which the specified key is mapped, or
+     *     {@code null} if this map contains no mapping for the key
+     * @throws NullPointerException if the specified key is null
+     */
+    public V getQuietly(Object key) {
+        final Node<K, V> node = data.get(key);
+        return (node == null) ? null : node.getValue();
+    }
+
+    @Override
+    public V put(K key, V value) {
+        return put(key, value, false);
+    }
+
+    @Override
+    public V putIfAbsent(K key, V value) {
+        return put(key, value, true);
+    }
+
+    /**
+     * Adds a node to the list and the data store. If an existing node is found,
+     * then its value is updated if allowed.
+     *
+     * @param key key with which the specified value is to be associated
+     * @param value value to be associated with the specified key
+     * @param onlyIfAbsent a write is performed only if the key is not already
+     *     associated with a value
+     * @return the prior value in the data store or null if no mapping was found
+     */
+    V put(K key, V value, boolean onlyIfAbsent) {
+        checkNotNull(key);
+        checkNotNull(value);
+
+        final int weight = weigher.weightOf(key, value);
+        final WeightedValue<V> weightedValue = new WeightedValue<V>(value, weight);
+        final Node<K, V> node = new Node<K, V>(key, weightedValue);
+
+        for (;;) {
+            final Node<K, V> prior = data.putIfAbsent(node.key, node);
+            if (prior == null) {
+                afterWrite(new AddTask(node, weight));
+                return null;
+            } else if (onlyIfAbsent) {
+                afterRead(prior);
+                return prior.getValue();
+            }
+            for (;;) {
+                final WeightedValue<V> oldWeightedValue = prior.get();
+                if (!oldWeightedValue.isAlive()) {
+                    break;
+                }
+
+                if (prior.compareAndSet(oldWeightedValue, weightedValue)) {
+                    final int weightedDifference = weight - oldWeightedValue.weight;
+                    if (weightedDifference == 0) {
+                        afterRead(prior);
+                    } else {
+                        afterWrite(new UpdateTask(prior, weightedDifference));
+                    }
+                    return oldWeightedValue.value;
+                }
+            }
+        }
+    }
+
+    @Override
```

Comment:
We are not going to remove anything, or clear the Map, ever; the only mechanism being expiration...
Although if simple enough could leave `clear()` I suppose.

Reply:
Sigh. That'd be unfortunate.

It sounds like maybe I should try out removing pieces until things fail, then putting back. :-)

Reply:
I don’t think a tiny custom cache is too bad. I’m not offended if you don’t use CLHM or anything, and we could write something very specific to your interface and iterate on the tests.

Reply:
I am bit torn here. I like the idea of reusing well-worn components, as well as "smallest thing that works well enough" (i.e. better than current sub-optimal approach).

I think use of CLHM makes sense still, if that is the shortest path. Let's try it out first?

Reply:
@pjfanning can we make the CLHM classes package-private? Then it is a hidden implementation detail.

Reply:
to clarify, I mean that the entire forked code could be under `util` instead of `util.clhm` and made package private. Then only `LRUMap` can access those classes and refine its implementation w/o leaking anything externally.

Reply:
I've just made a couple more classes package private - so only ConcurrentLinkedHashMap itself should be public - I also modified package-info.java to suggest that the package is for internal use and to use the official concurrentlinkedhashmap-lru if users want the use the map for their own purposes.

Reply:
@cowtowncoder what do you think about Ben's idea to move all the new classes to util package and making them all package private? I have a small preference for keeping them in a separate subpackage but am willing to go with the consensus.

Reply:
I think I like namespacing here, not keeping everything at `util` level (since there are other things).
In 2.x it would leave CHLM -- which we should actually rename to something else just to avoid simple auto-completion problem... PrivateCHLM? ShadedCHLM? JacksonCHLM? anything but not plain `ConcurrentLinkedHashMap`, basically -- open, `public` and theoretically usable. But as long as name doesn't match something users expect I think it avoids 90% of unintended use.

In 3.0 we have more freedom to move classes around.